packages/circuits/helpers/substring-match.circom

```diff
@@ -0,0 +1,134 @@
+pragma circom 2.1.6;
+
+include "circomlib/circuits/comparators.circom";
+include "circomlib/circuits/mux1.circom";
+include "../utils/hash.circom";
+
+/// @title SubstringMatch
+/// @notice This template verifies if a given substring exists within a larger string at a specified index
+/// @dev Uses a Random Linear Combination (RLC) approach to efficiently compare substrings
+/// @param maxLength The maximum length of the input string
+/// @param maxSubstringLength The maximum length of the substring to be matched
+/// @input in An array of ASCII values representing the input string
+/// @input startIndex The starting index of the substring in the input string
+/// @input revealedString An array of ASCII values representing the substring to be matched
+/// @input r A random value used for the RLC calculation
+/// @output isValid A signal that is 1 if the substring matches at the given index, 0 otherwise
+template SubstringMatch(maxLength, maxSubstringLength) {
```

Comment:
What if we VarLeftShift `in` by `startIndex` and then compare first `endIndex - startIndex` bytes of both `shiftedIn` and `revealedString`? (Or even simply reveal that as output, instead of revealString as an input). We can do this mostly with existing circuits like `PackByteSubarray`.

(Not sure which is more constraint efficient but something to compare if you are down)

Reply:
Thanks for pointing this out @saleel. The `SelectSubArray` template is indeed much more efficient than our current approach. 

`SelectSubArray`
<img width="338" alt="Screenshot 2024-08-21 at 09 06 24" src="https://github.com/user-attachments/assets/30efeded-7276-4575-8679-766213be74c8">

`SubstringMatch`
<img width="338" alt="Screenshot 2024-08-21 at 09 08 57" src="https://github.com/user-attachments/assets/0f1915e3-69fd-457d-8bc1-944ed62d86c9">

And I assume that it has been audited before too. I believe that using `SelectSubArray` is better for revealing substrings. We can delete this PR. 

cc: @Divide-By-0

Reply:
Nice. Thank you for the benchmarks.
Yes, if there is no explicit need for comparing values for equality in circuit we can simply use `SelectSubArray`.

Reply:
No, we just want to reveal the substring.

Reply:
@Divide-By-0  FYI ^ 
I think can close this PR if you also agree.

Reply:
@saleel made some changes, can you have a look?